packages/block-editor/src/components/iframe/index.js

```diff
@@ -0,0 +1,226 @@
+
+			if ( cancelled ) {
+				event.preventDefault();
+			}
+		}
+
+		const eventTypes = [ 'keydown', 'keypress', 'dragover' ];
+
+		for ( const name of eventTypes ) {
+			doc.addEventListener( name, bubbleEvent );
+		}
+
+		return () => {
+			for ( const name of eventTypes ) {
+				doc.removeEventListener( name, bubbleEvent );
+			}
+		};
+	}, [] );
+}
+
+/**
+ * Sets the document direction.
+ *
+ * Sets the `editor-styles-wrapper` class name on the body.
+ *
+ * Copies the `admin-color-*` class name to the body so that the admin color
+ * scheme applies to components in the iframe.
+ *
+ * @param {Document} doc Document to add class name to.
+ */
+function useBodyClassName( doc ) {
+	useEffect( () => {
+		doc.dir = document.dir;
+		doc.body.className = BODY_CLASS_NAME;
+
+		for ( const name of document.body.classList ) {
+			if ( name.startsWith( 'admin-color-' ) ) {
+				doc.body.classList.add( name );
+			}
+		}
+	}, [] );
+}
+
+/**
+ * Positions the body element so that the resize listener works correctly. We're
+ * using an absolute position here because the resize listener doesn't seem to
+ * report shrinking when the position is relative, causing the iframe not to
+ * shrink when content is removed.
+ *
+ * @see https://github.com/FezVrasta/react-resize-aware#usage
+ *
+ * @param {Document} doc Document to set styles to.
+ */
+function useResizeListenerCompat( doc ) {
+	useEffect( () => {
+		// Necessary for the resize listener to work correctly.
+		doc.body.style.position = 'absolute';
+		doc.body.style.right = '0';
+		doc.body.style.left = '0';
+	}, [] );
+}
+
+/**
+ * Sets the document head and default styles.
+ *
+ * @param {Document} doc  Document to set the head for.
+ * @param {string}   head HTML to set as the head.
+ */
+function useHead( doc, head ) {
+	useEffect( () => {
+		doc.head.innerHTML =
+			// Body margin must be overridable by themes.
+			'<style>body{margin:0}</style>' +
+			'<style>.wp-block[data-align="full"],.wp-block.alignfull{max-width:100vw!important;width:100vw!important;}</style>' +
+			head;
+	}, [] );
+}
+
+function IframeContent( { doc, head, children } ) {
+	useHead( doc, head );
+	useStyleSheetsCompat( doc );
+	useBubbleEvents( doc );
+	useBodyClassName( doc );
+	useResizeListenerCompat( doc );
+	return createPortal( children, doc.body );
+}
+
+export default function Iframe( { children, head, style = {}, ...props } ) {
+	const [ resizeListener, sizes ] = useResizeObserver();
+	const [ contentDocument, setContentDocument ] = useState();
+	const ref = useRef();
+
+	function setDocumentIfReady( doc ) {
+		const { readyState } = doc;
+
+		if ( readyState === 'interactive' || readyState === 'complete' ) {
+			setContentDocument( doc );
+		}
+	}
+
+	useEffect( () => {
+		setDocumentIfReady( ref.current.contentDocument );
+	}, [] );
+
+	function setRef( newRef ) {
+		ref.current = newRef;
+
+		if ( newRef ) {
+			setDocumentIfReady( newRef.contentDocument );
```

Comment:
It seems this is called three times to catch different flows/edge cases, I wonder if it would be good to abstract in a hook and I wonder if we should be using `domReady` somehow here?

Reply:
I simplified this now. I don't think that `domReady` is good here since the goal is different. For the iframe, we're looking for a usable document as quickly as possible. Only Firefox is a bit slow here, where it's only ready on the `load` event. In other browsers the document is ready as soon as the iframe mounts.